std/hash_map.zig

```diff
@@ -135,15 +135,15 @@ pub fn HashMap(comptime K: type, comptime V: type, comptime hash: fn (key: K) u3
             return res.kv;
         }
 
-        fn ensureCapacity(self: *Self) !void {
-            if (self.entries.len == 0) {
-                return self.initCapacity(16);
+        /// Sets the capacity to the new capacity if the new
+        /// capacity is greater than the current capacity.
+        pub fn ensureCapacity(self: *Self, new_capacity: usize) !void {
+            if (new_capacity <= self.entries.len) {
+                return;
             }
-
-            // if we get too full (60%), double the capacity
```

Comment:
I'm having trouble thinking about when this would come into play. How could `ensureCapacity` be called with a value that would also keep the HashMap at 60% full (after the "round to the nearest power of two" is added)? Could you provide an example of the calls that would lead to bad performance?

I can think of this, if `putAssumeCapacity` is added:
```zig
// map is empty
map.ensureCapacity(8);
map.putAssumeCapacity(1);
map.putAssumeCapacity(2);
map.putAssumeCapacity(3);
map.putAssumeCapacity(4);
map.putAssumeCapacity(5);
// it's 60% full after the last put, but it won't be resized
map.putAssumeCapacity(6);
map.putAssumeCapacity(7);
```
but I'm not sure how moving the >= 60% full check into `ensureCapacity` would change that case.

Reply:
> I'm not sure how moving the >= 60% full check into `ensureCapacity` would change that case.

It would mean that if you `map.ensureCapacity(8)`, then it really allocates room for e.g. 16 elements.
That would mean that you don't get performance degradation as the hash table fills up.

Reply:
Ok, I think I see what you mean. Something like this scenario?

```zig
// map is empty
var i=0;
while (i < 9) : (i += 1) {
    map.put(i);
}
// capacity is 16, size is 9
// the next put would normally trigger a resize
map.putAssumeCapacity(10);
map.putAssumeCapacity(11);
// this will round up to 16, which would still leave the map
// >= 60% full, so it should actually get set to 32 capacity
map.ensureCapacity(15);
```

Reply:
```zig
map.ensureCapacity(1024)
var i=0;
while (i < 1024) : (i += 1) {
    map.putAssumeCapacity(i); // as the hash gets fuller; each put gets slower and slower
}
map.get(555); // likely to be *very* slow: worst case is having to traverse/check all 1024 elements!
```

And note that different hash table implementations have different ideal fill-rates.
e.g. some want to be <20% full. other implementations are happy to be 80% full.
To avoid having users of `std.HashMap` hard code assumptions about hash table behaviour, the standard interfaces should take care of it.

Reply:
Your example is still unavoidable, though, right? Or would `putAssumeCapacity` still try to resize and just swallow the error if it can't?

Reply:
> Your example is still unavoidable, though, right?

Howso? My example *should* work, but preferably we overallocate (up to 2096) so that the slowdown doesn't happen.

> Or would putAssumeCapacity still try to resize and just swallow the error if it can't?

oooo, interesting.....

------

I wonder if we need a second function: `.optimizeForCapacity(usize: capacity)`. It would be a way to also shrink the hash (as we don't currently have a function for that).

Reply:
> preferably we overallocate (up to 2096) so that the slowdown doesn't happen

Oh, you're saying just do an unconditional `* 2` of anything passed to ensureCapacity?

Reply:
> Oh, you're saying just do an unconditional `* 2` of anything passed to ensureCapacity?

Almost. `*5/3` and round up to next power of two.

Reply:
Let me know what you think of the changes in the latest commit. I made `ensureCapacity` optimize capacity as you suggested, and then added a `ensureCapacityExact` fn that just blindly sets the capacity to the passed in amount. Not sure if `ensureCapacityExact` needs to be public.

With some debug printing, this is what your example (the `ensureCapacity(1024)` -> `put` 1024 items example) looks like:
```
optimizing for expected_count: 1024
expected_count * 5 / 3 = 1706
rounded to power of two = 2048
initCapacity: 2048
```

If `ensureCapacityExact(1024)` is used instead, then it looks like:
```
initCapacity: 1024
initCapacity: 2048
```
(since the a `put` call resizes it to 2048 once it gets 60% full).

Reply:
I think it's better now (good enough to merge; could be improved in future).
I might suggest making `ensureCapacityExact` a private function until we are sure about it.

Any chance you could add `putAssumeCapacity` to this PR?

Reply:
> Any chance you could add `putAssumeCapacity` to this PR?

Doing that now. Should there also be `getOrPutAssumeCapacity` variants?

Reply:
> Should there also be `getOrPutAssumeCapacity` variants?

I think leave it out for now, `getOrPut` has issues of its own.